File: atomic-x/android/src/main/kotlin/io/trtc/tuikit/atomicx/albumpicker/AlbumPickerPlugin.kt
package io.trtc.tuikit.atomicx.albumpicker

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * AlbumPickerPlugin module
 *
 * AlbumPickerPlugin acts as an intermediary layer between AtomicXPlugin and AlbumPickerHandler, managing MethodChannels and EventChannels.
 *
 * Note: This is distinct from Android Compose's AlbumPicker; this class is specifically designed for the Flutter Plugin layer.
 */
class AlbumPickerPlugin(
    private val pluginBinding: FlutterPlugin.FlutterPluginBinding
) : MethodCallHandler, EventChannel.StreamHandler {

    companion object {
        private const val METHOD_CHANNEL_NAME = "atomic_x/album_picker"
        private const val EVENT_CHANNEL_NAME = "atomic_x/album_picker_events"
    }

    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private val albumPickerHandler = AlbumPickerHandler { event ->
        eventSink?.success(event)
    }

    init {
        methodChannel = MethodChannel(pluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel?.setMethodCallHandler(this)
        
        eventChannel = EventChannel(pluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel?.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "pickMedia" -> {
                albumPickerHandler.handlePickMedia(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    // EventChannel.StreamHandler
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        android.util.Log.d("AlbumPickerPlugin", "EventChannel onListen")
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        android.util.Log.d("AlbumPickerPlugin", "EventChannel onCancel")
        eventSink = null
    }

    fun dispose() {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        eventSink = null
    }
}

File: atomic-x/ios/Classes/AlbumPicker/AlbumPickerPlugin.swift
import Flutter
import UIKit

/**
 * AlbumPickerPlugin
 * 
 * AlbumPickerPlugin 作为 AtomicXPlugin 和 AlbumPickerHandler 之间的中间层，管理 MethodChannel 和 EventChannel
 * 
 * 注意：这与 iOS SwiftUI 的 AlbumPicker 不同；此类专门为 Flutter Plugin 层设计
 */
class AlbumPickerPlugin: NSObject, FlutterStreamHandler {
    private static let channelName = "atomic_x/album_picker"
    private static let eventChannelName = "atomic_x/album_picker_events"
    
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var albumPickerHandler: AlbumPickerHandler?
    private var eventSink: FlutterEventSink?
    
    init(registrar: FlutterPluginRegistrar, viewController: UIViewController?) {
        super.init()
        
        methodChannel = FlutterMethodChannel(
            name: AlbumPickerPlugin.channelName,
            binaryMessenger: registrar.messenger()
        )
        
        eventChannel = FlutterEventChannel(
            name: AlbumPickerPlugin.eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        
        eventChannel?.setStreamHandler(self)
        
        albumPickerHandler = AlbumPickerHandler(viewController: viewController, eventSink: { [weak self] event in
            self?.eventSink?(event)
        })
        
        methodChannel?.setMethodCallHandler { [weak self] (call, result) in
            self?.handleMethodCall(call, result: result)
        }
    }
    
    // MARK: - FlutterStreamHandler
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("[AlbumPickerPlugin] EventChannel onListen")
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("[AlbumPickerPlugin] EventChannel onCancel")
        self.eventSink = nil
        return nil
    }
    
    // MARK: - MethodChannel Handler
    
    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pickMedia":
            albumPickerHandler?.handlePickMedia(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func dispose() {
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
        eventChannel?.setStreamHandler(nil)
        eventChannel = nil
        albumPickerHandler = nil
        eventSink = nil
    }
}
